### DISCUSSION TITLE
Automatic On and Off at Sunset and Sunrise

### DISCUSSION THREAD

**Discussion by Misunderstood-Wookiee** [basnijholt/adaptive-lighting#419] [Ideas]
I have been using automation to call the services manually as needed for sunset and sunrise in my home however it would be far easier if the choice existed inside adaptive lighting to have an Automatic Off period for a configurable state of sun elevation with perhaps an offset value that gets used for cloudy and or rainy weathers ( as usually, it can be darker outside)

This can then be setup per room and have lights automatically turn off lights during sunrise if desired. 
It can also be interrogated into sleep mode service so at some time of the night sleep mode can automatically turn on for rooms that have this enabled.

Just some thoughts to help make less work of custom automation for basic lighting when there are plugins like this which could be an All in one solution and just modified there.

**Comment by th3w1zard1** [comment]
IMO there's already too many config options as it is. Is it not enough to be able to add an automation trigger from hass when the sun rises/sets?

![image](https://user-images.githubusercontent.com/2219836/226812743-65d1ad6f-cff2-41c0-a8b3-66bad15aa38a.png)

**Reply by Misunderstood-Wookiee** [reply]
Sure, while that does technically work and I had been doing just that until a few months back it's not as accurate as the numerical state of elevation property of the sun entity due to seasonal changes the sunset and sunrise can be way off and offsets don't cover seasonal change so as seasonal changes you'll find the time difference fluctuates. 

This was suggested to me by other senior members of the HASS g
Discord I also then tried to make automation to get weather and adjust accordingly as I found rain or overcast/Cloudy days would make the house much too dark earlier and would need lights to come on sooner. 

So I had an offset for weather events. 

Which wasn't quite working but almost configured I decided to revert though for the hassle was a bit much for the automation routine. 

And I thought perhaps something simple could be done here to account for elevation value and weather types as an integration AIO solution than make an automation routine at all this way the integration is more than capable of correctly handling the best time of day to come off and on and adjust for forecasted weather events to some degree based on how accurate the hass servers information is of choosing weather integration.

**Reply by Misunderstood-Wookiee** [reply]
The integration methods for getting sun position could be adjusted to also read the servers home location elevation value (which users should manually set once)  when passed to a param as mentioned provides near perfect real world sun pos across entire year for that exact location and altitude the only thing missing then is the ability for telling it at what elevation above or below horizon should a room have the lights on or off then we are just missing knowing it's if it's overcast essentially to offset this by some time value to account for lesser lighting conditions. 

I dont know how much work it would be to add or re work code as I have no experience with the current code base or HASS coding. 

I wad just offering an idea here that could elevate the integration into a more robust solution that takes care of lights per room without having to rely on automations to override things.

**Reply by th3w1zard1** [reply]
From https://astral.readthedocs.io/en/latest/#effect-of-elevation

##  Effect of Elevation
### Times Of The Sun
> The times of the sun that you experience depend on what obscurs your view of it. It may either be obscured by the horizon or some other geographical feature (e.g. mountains)
> 
> If what obscures you at ground level is the horizon and you are at a elevation above ground level then the times of the sun depends on how far further round the earth you can see due to your elevation (the sun rises earlier and sets later).
> 
> The extra angle you can see round the earth is determined by calculating the angle α in the image below based on your elevation above ground level, and adding this to the depression angle for the sun calculations.
> 
> ![image](https://user-images.githubusercontent.com/2219836/226839832-10872d72-7984-407e-8829-65a28c97d1ee.png)
> If your view is obscured by some other geographical feature than the horizon, then the adjustment angle is based on how far you are above or below the feature and your distance to it.
> 
> For the first case i.e. obscured by the horizon you need to pass a single float to the Observer as its elevation. For the second case pass a tuple of 2 floats. The first being the vertical distance to the top of the feature and the second the horizontal distance to the feature.
> 
 ### Elevation Of The Sun
> Even though an observer’s elevation can significantly affect the times of the sun the same is not true for the elevation angle from the observer to the sun.
> 
> As an example the diagram below shows the difference in angle between an observer at ground level and one on the ISS orbiting 408 km above the earth.
> 
> ![image](https://user-images.githubusercontent.com/2219836/226840030-9d9bad39-3fae-45f6-b910-e3b9fefbcbca.png)
> The largest difference between the two angles is when the angle at ground level is 1 degree. The difference then is approximately 0.15 degrees.
> 
> At the summit of mount Everest (8,848 m) the maximum difference is 0.00338821 degrees.
> 
> Due to the very small difference the astral package does not currently adjust the solar elevation for changes in observer elevation.

### Effect of Refraction
> When viewing the sun the position you see it at is different from its actual position due to the effect of atmospheric [refraction](https://en.wikipedia.org/wiki/Refraction) which makes the sun appear to be higher in the sky. The calculations in the package take this refraction into account.
> 
> The [sunrise()](https://astral.readthedocs.io/en/latest/package.html#astral.sun.sunrise) and [sunset()](https://astral.readthedocs.io/en/latest/package.html#astral.sun.sunset) functions use the refraction at an angle when the sun is half of its apparent diameter below the horizon. This is between about 30 and 32 arcminutes and for the astral package a value of 32” is used.

**Reply by th3w1zard1** [reply]
From HASS:
```
@callback
@bind_hass
def get_astral_location(
    hass: HomeAssistant,
) -> tuple[astral.location.Location, astral.Elevation]:
    """Get an astral location for the current Home Assistant configuration."""

@callback
def get_location_astral_event_next(
    location: astral.Location,
    location: astral.location.Location,
    elevation: astral.Elevation,
    event: str,
    utc_point_in_time: datetime.datetime | None = None,
    offset: datetime.timedelta | None = None,
) -> datetime.datetime:
    """Calculate the next specified solar event."""
```

**Reply by th3w1zard1** [reply]
Adaptive-lighting calls get_astral_location() with no arguments. If we added a config option to include elevation, would that work for you?

EDIT: adaptive-lighting doesn't even use the elevation! See this code:
```
        _loc = get_astral_location(self.hass)
        if isinstance(_loc, tuple):
            # Astral v2.2
            location, _ = _loc
```

**Reply by th3w1zard1** [reply]
We already considered weather types in another issue awhile back, while it seemed promising it didn't affect the overall color temperature of the light in any significant way. See https://github.com/basnijholt/adaptive-lighting/issues/164#issuecomment-1150412962

**Reply by Misunderstood-Wookiee** [reply]
> Adaptive-lighting calls get_astral_location() with no arguments. If we added a config option to include elevation, would that work for you?
> 
> EDIT: adaptive-lighting doesn't even use the elevation! See this code:
> ```
>         _loc = get_astral_location(self.hass)
>         if isinstance(_loc, tuple):
>             # Astral v2.2
>             location, _ = _loc
> ```

Thatsq sounds about right, the hass does already for some time now have home location taking elevation which when set by user on initial or post setup gives you the values you speak of for sun calculations. This is how it's possible to male automation using the numeric_state of sun that is more accurate thrb just time of day and location alone. 


So yes if adaptive lighing could leverage this it would make its automation service overall more reliable in that sense of getting a more accurate sun position  for everyone.

**Reply by Misunderstood-Wookiee** [reply]
> We already considered weather types in another issue awhile back, while it seemed promising it didn't affect the overall color temperature of the light in any significant way. See https://github.com/basnijholt/adaptive-lighting/issues/164#issuecomment-1150412962

Ah see I wasn't planning on using it to effect colour rather it know its going to be dark overall and maybe light the home up a wee bit on rooms with the option enabled for say rooms where it already lacks good sun light as is case with hallways
